Add TimeDelta format for Year Day Hour Minute Sec string

taldcroft opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->

This is a draft pull request to add a new string `TimeDelta` format that represents the time delta as a string in format that is analogous to the Time `ymdhms` format. The driver for this is that it can be convenient to have a human-readable string representing a time delta which has natural scales so that delta from seconds to years are manageable.

This PR defines a new format which is like `"[-+]? 1yr 2d 3hr 4min 5.6s"`.

The only other existing standard that I am aware of for time intevals is defined in ISO 8601 durations that represent the amount of time between two dates or times. The ISO 8601 duration format is: P(n)Y(n)M(n)DT(n)H(n)M(n)S. The fundamental problem with this format is that it explicitly depends on the reference start date as a calendar date in Year Month day hour min sec. So this is not suitable for physical time deltas that are independent of the reference dat.e

This PR is a WIP but before going much further I wanted to get feedback on the main idea of defining and adding a new format. If that is OK, then discussion of details of the format and implementation details would follow.

Especially interested in thoughts from @mhvk.

Known issues:
- [ ] Full precision is likely not preserved.
- [ ] Warning when passing a float time delta without a unit is not emitted.

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

<!-- Optional opt-out -->

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

mhvk reviewed (commented):
I think this is a tricky format. It is definitely most logical to use a year of 365.25 days (note the docstring states 365 or 366!). But will users really expect the following?
```
In [4]: td = TimeDelta(np.arange(364, 368), format='jd')

In [5]: td.ydhms
Out[5]: array(['+364d', '+365d', '+1yr 18hr', '+1yr 1d 18hr'], dtype='<U12')
```
I noticed that `datetime.timedelta` does not have years (or months), perhaps for this reason (it does have weeks). Though that format is pretty weird, with the sign applying just to the days, so one gets
```
In [9]: print(timedelta(hours=-6))
-1 day, 18:00:00
```
Anyway, that format we have already:
```
In [13]: np.vectorize(str)(td.datetime)
Out[13]: 
array(['364 days, 0:00:00', '365 days, 0:00:00', '366 days, 0:00:00',
       '367 days, 0:00:00'], dtype='<U17')
```

Overall, my tendency would be to not use years at all (at least not by default). I think I'd also shorten the notation, to use `01h02m03.04s` for the time at least (similar to what is done for RA), and maybe just prepend `sNNNd`.

Also, I wonder a bit about the name: `TimeYMDHMS` returns a `recarray` - that would suggest `TimeDeltaYDHMS` should do the same. Though then the sign becomes annoying, like for angle tuples we just deprecated... In analogy with `TimeYearDay` (format `yday`), maybe `TimeDeltaYearDay`?

Anyway, not really sure what is the right solution! I do think an unambiguous string format would be nice to have, but wish there was a sensible standard we could simply implement, rather than rolling our own.

p.s. Looking a bit further, I found https://learn.microsoft.com/en-us/dotnet/standard/base-types/standard-timespan-format-strings - so, microsoft uses `±ddd.hh:mm:ss.ffff`. Rather weird...

taldcroft commented:
@mhvk - interesting about the microsoft format. It just seems pretty unreadable, and I don't understand why they made the different flavors. Especially one with a period after the days and one with a colon after days. This format also does not formally support time deltas to astropy precision since it is limited to microsec.

About not supporting year and just use days. My thinking there was that year is a pretty common time unit in astronomy. Imagine a config file for some simulation that needs a time scale in years.

One alternate idea is to simply support any single float and astropy unit of time with space being optional, so e.g. `1e8yr`, `2d`, `3 wk`, `3.1312311 wk`, `1e18s`. The main question would be the output representation. This could be controlled by `out_subfmt` to specify a unit, or by default choose the largest unit where the number is greater than 1.0 (or something like that?). This is certainly simpler and doesn't feel like inventing a new format.

mhvk commented:
A single unit string would be a possibility. Right now we have
```
In [6]: td.to(u.s).to_string()
Out[6]: '[31449600. 31536000. 31622400. 31708800.] s'

In [7]: td.to(u.yr).to_string()
Out[7]: '[0.99657769 0.99931554 1.00205339 1.00479124] yr'
```
But this is not an array of string. Possibly, `Quantity` should gain that option... Especially since `Angle` does have it (an inconsistency that has long annoyed me...). So, right now one would have to do,
```
In [9]: [str(_t) for _t in td.to(u.yr)]
Out[9]: 
['0.9965776865160848 yr',
 '0.999315537303217 yr',
 '1.002053388090349 yr',
 '1.0047912388774811 yr']
```
Of course, this would still loose precision, so I think it makes sense to just expand our long-string  ability from `TimeNumeric`, which already works for `format='jd'` and `'sec'`:
```
In [17]: td = TimeDelta(np.arange(364, 368), 1e-10, format='jd')

In [18]: td.to_value('jd', subfmt='str')
Out[18]: 
array(['364.0000000001', '365.0000000001', '366.0000000001',
       '367.0000000001'], dtype='<U14')
```

One option would be to recognize `td.to_value(unit, subfmt='str')` (this currently errors, so we are free to adjust it). Not passing in a unit with `subfmt='str'` could in principle use a "logical" unit is chosen. Or one could put in `u.physical.time` for as a signal that a "good" unit should be chosen.

Although a proper format would make round-tripping easier. The only annoyance then is that somehow the format instance has to carry the unit. As you note, `subfmt` might be good for that, but that then looses the option to use it to select str, bytes, longdouble, etc.

taldcroft commented:
@mhvk - it seems your focus has been on leveraging existing pathways for generating string representations of a `TimeDelta`, but for me the primary driver is instantiating a `TimeDelta` from a string. For that I think that a new format is the only way.

For me this is really a string format so I don't think `TimeNumeric` has any applicability here. Allows `bytes` input is fine, but I think if the user needs `bytes` output they can do that themselves.

Fundamentally I think the key decision is having a multi-component format or something that looks basically like the string repr of Quantity.

### Multi-component like `-3d2h14m12.232342s` (with format details TBD)
- In this case `precision` would apply to the seconds component by default and so precision to 1e-9 sec is possible for arbitrarily long durations.
- More complicated.
- It's really a new format "standard" for time delta representation, though with plenty of precedent in astronomy.
- The format you suggested with h, m, s is potentially ambiguous with an hour angle (if there is no other context). That's part of why I originally suggested the longer format with the labels being exactly the astropy unit strings.
- `out_subfmt=<unit>` could force output in a single unit, with precision applying to that unit.
- No need to infer `out_subfmt` based on the input, which seems cleaner to me.
- Setting `out_subfmt="yr"` and `precision=5` would give an output like `12.12345y` (or `12.12345 yr` if we used the longer unit string labels, thus giving an output that looks like a Quantity str.
- We could also define a special `out_subfmt` that uses a single unit but chooses according to the same rule that the Quantity-like 

### Quantity-like `12.12345 yr`
- Simpler.
- Strictly speaking this would be a new format "standard", but the overlap with Quantity representation makes it seem like not really new.
- Precision would apply to the unit, so `precision=5` could give `12.12345 yr`. 
- It's a little messy with the `out_subfmt`. Imagine `TimeDelta(["12.12345 yr", "0.5 s"])`. Probably the answer is to infer the longest applicable unit but whatever choice we make will be a bit arbitrary and may surprise *someone*.

All in all I think multi-component is more flexible since it includes Quantity-like as a option, but there is always an argument for "simple is better than complex". Thoughts?

taldcroft commented:
I think if we make this format explicitly tied to astropy `Quantity` and the corresponding unit names then there is no ambiguity about the year definition even for the multi-component form. We are already accepting that implicit definition for the `Quantity-like` version `12.2 yr`. 

Likewise I don't really think the sign issue is a blocker for astronomers who all understand how to read -12d 30m 42.5s.

Having said that, I'm willing to go with `Quantity-like`, but only if we are agreed on not doing the multi-component later. I think it would be a mistake to patch `Quantity-like` later to handle multi-component with a sub-format since that would be messy [1]. Instead if we start with multi-component now then `Quantity-like` is a natural and simple sub-format of multicomponent (restricting to just one possible unit instead of all of them).

[1] I would the format name might even be `quantity` since it would be natural and nice to exactly share the string repr. Expanding to multi-component just wouldn't work. Also, "sub-format" really implies narrowing the format, not expanding it.

taldcroft commented:
And this might be the right point to invite comment on astropy-dev. I had expected this might be controversial.

mhvk commented:
@taldcroft - I'm not sure I see the problem with possible later expansion, but will admit I have not thought this through very well (am on holidays and just occasionally checking what's new). I think it is a good idea to ask at astropy-dev - in the end, we should be driven by the use cases people have!

taldcroft commented:
> I'm not sure I see the problem with possible later expansion

Maybe I'm being a bit pedantic, but `-3d 2hr 14min 12.232342s` does not seem like a sub-format of a format like `3.12312 d`. It feels more like a super-format or something entirely new.

mhvk commented:
I guess I saw it as effectively something we discussed for `Angles` too, where one has a "format string" that includes options like "hms".  But let's see whether we get input on what is preferred first!

taldcroft commented:
@mhvk - having received no input from the astropy-dev solicitation, I am pressing ahead with the full-featured implementation to see how it would look. The code is in a better state now and I have uploaded a notebook that I am using to play around with this:
https://gist.github.com/taldcroft/8ef284f25fc8671bfd878793310f0bb8

You'll notice some creep into core functionality changes (you know how that goes). We can always separate those out into separate PR's if needed.

taldcroft commented:
@mhvk - this is ready for review. I've updated the PR description, added tests, and fixed some initial issues. The only thing obviously missing are the What's New and change log.

taldcroft commented:
BTW the Python 3.9 failure looks to be unrelated.

pllim commented:
SAMP timeout expired failure in oldest deps job is indeed unrelated. I think it is transient.

mhvk commented on astropy/time/formats.py:
```diff
@@ -2122,7 +2123,7 @@ class TimeDeltaSec(TimeDeltaNumeric):
 
 
-class TimeDeltaJD(TimeDeltaNumeric):
+class TimeDeltaJD(TimeDeltaNumeric, TimeUnique):
```
This is not unique, right? I don't see how one can distinguish from `TimeDeltaSec`. Or is it that currently we treat the default as "day"?

taldcroft replied:
Yes, initializing `TimeDelta` with a plain number uniquely identifies that as JD because of the default as day. This works even for a string that represents a number e.g. `TimeDelta('12')`.

mhvk replied:
Would still be nice to deprecate that... But probably far too late!

mhvk commented on astropy/time/formats.py:
```diff
@@ -2175,6 +2176,204 @@ def value(self):
+        )
+
+        for jd1, jd2, out in iterator:
+            jd1_, jd2_ = day_frac(jd1, jd2)  # Why is this required?
```
I agree this should not be necessary.

taldcroft replied:
Removed.

mhvk commented on astropy/time/formats.py:
```diff
@@ -2175,6 +2176,204 @@ def value(self):
+    name = "quantity_str"
+
+    subfmts = (
+        ("multi", None, None),
```
I still do not like having years included by default - just too confusing - this is going to cause misery!

How about having "dhms" and "ydhms" options? (where I'd put "dhms" first on purpose, i.e., one has to do more work to get the years).

taldcroft replied:
Resolved by opening [#15482](https://github.com/astropy/astropy/issues/15482)

mhvk commented on astropy/time/formats.py:
```diff
@@ -2175,6 +2176,204 @@ def value(self):
+            ) = self.parse_string(val)
+
+        yrs, days, hrs, mins, secs = iterator.operands[1:]
+        jd1 = yrs * 365.25 + days + hrs / 24.0 + mins / 1440.0 + secs / 86400.0
```
There is no reason to insist on precision loss. Since `jd1` is days anyway, just do
```
jd1 = yrs * 365.25 + days  # Exact in usual case that yrs and days are integer
jd2 = hrs / 24.0 + mins / 1440.0 + secs / 86400.0  # Inexact
self.jd1, self.jd2 = day_frac(jd1, jd2)
```

taldcroft replied:
This reduces precision loss in common cases, but `12314343.234234234234232yr` is a valid input, or even `12314343.234234234234232yr 0.1231231231212312d 1231231.12312312312s`. I don't think the latter forms are converted without loss, so we cannot guarantee no loss.

mhvk replied:
Agreed, I just thought one might as well do the common case.

Though thinking about those longer strings, we do have the machinery for that: you can use `utils.decimal_to_twoval1` (this deals with strings too) and then `day_frac(v1, v2, factor=..., divisor=...)`. Obviously, this would only round-trip if one were to also do the reverse in output, with `twoval_to_string1`.

mhvk commented on astropy/time/formats.py:
```diff
@@ -2175,6 +2176,204 @@ def value(self):
+            jd = np.abs(jd)
+
+            if subfmt in ["*", "multi"]:
+                comps = self.get_multi_comps(jd)
```
I think one might as well keep precision here too, so `self.get_multi_comps(jd1_, jd2_)`

That also means above
```
if jd < 0:
    jd1_ = -jd1_
    jd2_ = -jd2_
    jd = np.abs(jd)
    sign = "-"
else:
    sign = ""
```

taldcroft replied:
Done.

mhvk commented on astropy/time/formats.py:
```diff
@@ -2175,6 +2176,204 @@ def value(self):
+        comp_names = ("yr", "d", "hr", "min", "s")
+        comp_scales = (365.25, 1.0, 1.0 / 24.0, 1.0 / 1440.0)
+        comps = []
+        for name, scale in zip(comp_names, comp_scales):
```
To keep precision:
```
jd, remainder = two_sum(jd1, jd2)
for name, scale in zip(comp_names, comp_scales):
    comp, jd = divmod(jd, scale)
    if name == "d":
        jd += remainder
...
```

mhvk commented on astropy/time/formats.py:
```diff
@@ -2175,6 +2176,204 @@ def value(self):
+
+        return self.mask_if_needed(np.array(iterator.operands[-1], dtype="U"))
+
+    def get_multi_comps(self, jd):
```
This could be quite easily adjusted to deal with a no-year multi-format:
```
def get_multi_comps(self, jd1, jd2, submt):  # assumed ydhms or dhms
# and then in the loop
     if name[0] not in subfmt:
        continue
```

taldcroft replied:
Defer for later discussion.

mhvk commented on astropy/time/tests/test_basic.py:
```diff
@@ -2361,9 +2361,10 @@ def test_hash_time_delta():
-    with pytest.raises(ValueError) as err:
+    with pytest.raises(
+        ValueError, match=r"Input values did not match any of the formats"
+    ) as err:
```
No need for "as err" any more

mhvk commented on astropy/time/tests/test_delta.py:
```diff
@@ -685,3 +685,100 @@ def test_no_units_warning():
+        "2yr 186d 7hr 35min 36.789s",
+        2.5 * 365.25 * 86400 + 3.5 * 86400 + 4.5 * 3600 + 5.5 * 60 + 6.789,
+    ),
+]
```
Maybe add a test for high precision? But that needs input of both jd1 and jd2 for the test.

mhvk commented on docs/time/index.rst:
```diff
@@ -1273,6 +1286,8 @@ name and optional `subformat`_ or a valid ``astropy`` unit::
 
 .. EXAMPLE END
 
+
```
Why the extra lines?

mhvk reviewed (commented):
I'm still not sold on including years in the multi-format, and feel strongly the default should be days, hours, minutes, and seconds. I suggest inline to have "dhms" for this as a subfmt. I'm OK, though, with adding "ydhms" as a non-default format - for people who understand what they are doing, and that 1yr = 365.25 d - this will *not* be what most people think when they see it (and they won't read the manual to check).

Otherwise, this looks all pretty good, and my main comments are that one might as well try to keep precision, as it is relatively trivial to do so!

taldcroft commented:
> I'm still not sold on including years in the multi-format

Let me try to sell you on this. :smile:

> 1yr = 365.25 d - this will not be what most people think when they see it.

Do you have data to back this assertion? I think the opposite, that most astronomers would pick 365.25 as the mean number of days in a year for doing calculations (but I also have no data). This format (via the name "quantity_str") explicitly makes the connection to astropy `Quantity` where indeed `yr` is `365.25d`.  Has there been feedback that this is a problem for `Quantity`?

>  (and they won't read the manual to check).

Even though I understand the realities, people will have to discover this format somehow and the year to days conversion will be clear from that. We can ensure that sphinx doc examples highlight this case, so that any search result shows this behavior.

> I suggest inline to have "dhms" for this as a subfmt. I'm OK, though, with adding "ydhms" as a non-default format.

I think this adds complexity and confusion.
- Code becomes more complex.
- Why do we allow "ydmhs" and "dhms" but not "hms", and "ms"?
- Explicitly enumerating the expected components "d", "h", "m", and "s" implies that they should all be in the output, bringing ambiguity to the format. Is it "1d 0h 0m 10.0s" or "1d 10.0s"?
- Users like me would wonder why "dhms" is first. Seems arbitrary.

This would definitely surprise me:
```
>>> TimeDelta("1yr").value
'365d 6hr'
```
Given that this format supports `yr` as a component, I would expect to get back `'1yr'`. So we have to carefully explain that  year special-cased because we think it would be too confusing.

mhvk commented on astropy/time/formats.py:
```diff
@@ -2175,6 +2176,204 @@ def value(self):
+                comp = int(comp)
+                comps.append(f"{comp}{name}")
+
+        sec = np.round(jd * 86400.0, self.precision)
```
Not directly related to the larger question, but this rounding at the end means you can wind up with `60.0s` - do you want to special-case that? As is,
```
TimeDelta(365.9999999999, format='jd').quantity_str
# '1yr 17hr 59min 60.0s'
```

To avoid duplication, maybe consider using some of the code in `astropy/coordinates/angles/formats.py`; as is,
```
from astropy.coordinates.angles import formats
formats.hours_to_string(.99999999999*24, precision=3, sep=['hr ', 'min ', 's'])
# '24hr 00min 00.000s'
```
I think the rounding bit could be quite easily factored out of `sexagesimal_to_string`.

taldcroft replied:
I was thinking about that corner case this morning and agree this should be done right. Thanks for the heads-up on existing code, I'll have a look.

taldcroft replied:
Fixed

mhvk commented:
Hmm, you're last example certainly makes a good point... Perhaps I am too worried (though I so seem to recall us having had some issues raised about it before).

But the note about omitting zeros also makes me wonder: is that generically what would be wanted? For a table of this, I think I'd rather see all of the entries, as it is easier to compare (like we do with angles). In that sense, there is perhaps a case for `dhms` (and `ydhms`) as you interpreted it: give me the value in (year +) day + sexagesimal.

taldcroft commented:
@mhvk - the idea of making this like sexagesimal opens up a new question of `1yr 2d 3hr 0min 0.0s` (minimal) or `1yr 002d 03h 00min 00.000s` (fixed format)? Only the latter fixed format version would format nicely in a table column.

All in all, I think it's worth remembering that the biggest driver for this format is a way to *input* time deltas as a string in a flexible and readable format. 

For the output since there are questions, maybe we could mark this feature API as experimental and see if there is any feedback from the community in the first release?

mhvk commented:
Yes, seems fair to see what people actually want! (My thought was indeed the fixed-format case.)

mhvk commented on astropy/time/core.py:
```diff
@@ -561,12 +562,13 @@ def _get_time_fmt(self, val, val2, format, scale, precision, in_subfmt, out_subf
+        if format is None:
+            # If val and val2 broadcasted shape is () then we cannot guess format
+            # from the input values.  Instead use the default format.
+            if val.size == 0 and (val2 is None or val2.size == 0):
```
Ah, in that case, do update the comment to say broadcasted shape `(0,)` - it was the comment that threw me off!

taldcroft commented:
@mhvk - after all that discussion I discovered a more fundamental problem with including years in the output format. The year component is an integer, but it should roll over at 365.25 or "365d 6hr" (which would indeed be very confusing). I found this when fixing the overflow for times with "60.0s" as the last component. So "364d 23hr 59min 60.0s" needs to go to "365d", but the normal rule would be to roll to "1yr", but that's wrong.

So `yr` is still allowed as an input unit, but it is not part of the default `multi` subformat.

taldcroft commented:
With that can we un-experimental this feature?

mhvk commented:
Sounds reasonable. Note that I'm stuck writing my next 5-year research grant (due Nov 1st) and will have very limited time for review. I think this was in good shape already, and am happy to try to review, but I better hold off looking further until you ping me explicitly.

taldcroft commented:
@mhvk - as far as I can see I have addressed all the review comments, fixed coverage issues, etc. The CI just started however so I need to see if that is good.

taldcroft commented:
BTW I definitely prefer to have the commits in this PR squashed. This has wandered a bit and there are a lot of small commits with little value on their own.

mhvk commented on docs/time/index.rst:
```diff
@@ -1257,11 +1261,22 @@ Use of the |TimeDelta| object is illustrated in the few examples below::
   >>> t1 + 1 * u.hour
   <Time object: scale='utc' format='iso' value=2010-01-01 01:00:00.000>
 
+  # Human-readable multi-scale format for string representation of a time delta.
```
Can we take the comment out of the block so there can be a proper link to `TimeDeltaQuantityStr`?

taldcroft replied:
Done.

mhvk reviewed (approved):
This looks good. I think we could keep precision for the single-unit case fairly easily but think it is fine to postpone that to follow-up, so I'll approve now.

taldcroft commented on astropy/time/formats.py:
```diff
@@ -2086,6 +2098,9 @@ class TimeDeltaFormat(TimeFormat):
 
     _registry = TIME_DELTA_FORMATS
+    _default_precision = 3
+    _min_precision = -99
```
Agreed the 99 is weird and I struggled a bit there. For `Time` there is a real limit to precision because that gets passed into ERFA routines. For `TimeDelta` there is not really a limit. Currently for `TimeDelta` formats the precision only gets used via `np.round()` in `quantity_str`.

There are other options like overriding the precision setter or allowing `None` as a value for `_min/max_precision`, but I think keeps the code simple.

mhvk replied:
OK, let's keep it as is.

taldcroft commented on astropy/time/formats.py:
```diff
@@ -2175,6 +2190,229 @@ def value(self):
+
+    def parse_string(self, timestr):
+        """Read time from a single string"""
+        # Datetime components required for conversion to JD by ERFA, along
```
Fixed

taldcroft commented on astropy/time/formats.py:
```diff
@@ -2175,6 +2190,229 @@ def value(self):
+                comps = self.get_multi_comps(jd1, jd2)
+
+            else:
+                value = (jd * u.day).to_value(subfmt)
```
Another day. :smile:

pllim commented:
Is this merge-able?

pllim commented:
I think the codecov upload failed. It can be flaky. Don't worry about it.
